### PR TITLE
Add diphone factored hybrid model output block

### PR DESCRIPTION
**Pull request by NeoLegends**
No triphone yet as there isn't an ONNX-compatible factored hybrid feature scorer yet and the diphone is compatible w/ `PrecomputedHybridFeatureScorer`. Support can be trivially amended though.

### CODE REVIEW

File: i6_models/parts/factored_hybrid/util.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+from enum import Enum
+from typing import Callable, Union
+
+from torch import nn
+
+
+class PhonemeStateClassV1(Enum):
```
Comment (michelwi) [inline i6_models/parts/factored_hybrid/util.py:7]:
```suggestion
class PhonemeBoundaryClassV1(Enum):
```
"state" to me sounds like it would be related to `num_hmm_states_per_phone`

File: i6_models/parts/factored_hybrid/util.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+) -> int:
+    """
+    :return: number of center phonemes given the augmentation values
+    """
+
+    factor = ph_class.factor() if isinstance(ph_class, PhonemeStateClassV1) else ph_class
+    return n_contexts * num_hmm_states_per_phone * factor
+
+
+def get_mlp(
```
Comment (michelwi) [inline i6_models/parts/factored_hybrid/util.py:42]: There is also https://github.com/rwth-i6/i6_models/blob/main/i6_models/parts/ffnn.py that one could reuse. But some linear layers seems simple enough to me that I do not care about code duplication here.

File: i6_models/parts/factored_hybrid/diphone.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            center_logits = self.center_encoder(center_features)  # C, B, T, F'
+            center_probs = F.log_softmax(center_logits, -1)
+            center_probs = center_probs.permute((1, 2, 3, 0))  # B, T, F', C
+            left_probs = F.log_softmax(left_logits, -1)
+            left_probs = left_probs.unsqueeze(-2)  # B, T, 1, C
+
+            joint_probs = center_probs + left_probs  # B, T, F', C
+            joint_probs = torch.flatten(joint_probs, start_dim=2)  # B, T, joint
+
+            return DiphoneProbsV1(
```
Comment (michelwi) [inline i6_models/parts/factored_hybrid/diphone.py:163]: In the full diphone case we do not need the `output_left` or the embeddings, do we?

Would it make sense to use a different Dataclass here?

File: i6_models/parts/factored_hybrid/diphone.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+            num_layers=cfg.context_mix_mlp_num_layers,
+            dropout=cfg.dropout,
+            activation=cfg.activation,
+        )
+
+    def forward(
+        self,
+        features: Tensor,  # B, T, F
+        contexts_left: Optional[Tensor] = None,  # B, T
+    ) -> Union[DiphoneLogitsV1, DiphoneProbsV1]:
```
Comment (michelwi) [inline i6_models/parts/factored_hybrid/diphone.py:113]: I know what you semantically want to say here, but since DiphoneProbsV1 inherits from DiphoneLogitsV1, programmatically I think this would be equivalent to
```suggestion
    ) -> DiphoneLogitsV1:
```

This change directly contradicts one other comment I made below, pls apply only one of those :P

File: i6_models/parts/factored_hybrid/diphone.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        phoneme_state_class: the phoneme state augmentation to apply
+        activation: activation function to use in the context mixing MLP.
+    """
+
+    left_context_embedding_dim: int
+    n_contexts: int
+    num_hmm_states_per_phone: int
+    phoneme_state_class: Union[int, PhonemeStateClassV1]
+
+    activation: Callable[[], nn.Module]
```
Comment (michelwi) [inline i6_models/parts/factored_hybrid/diphone.py:60]: would that be a function that takes no input but returns a Module?

In other places I have seen the typing
```suggestion
    activation: Union[nn.Module, Callable[[torch.Tensor], torch.Tensor]]
```
which seems more appropriate to me

File: i6_models/parts/factored_hybrid/diphone.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+            # here we forward every context to compute p(c, l|x) = p(c|l, x) * p(l|x)
+
+            features = features.expand((self.n_contexts, -1, -1, -1))  # C, B, T, F
+            embedded_left_contexts_ = embedded_left_contexts.reshape((self.n_contexts, 1, 1, -1)).expand(
+                (-1, *features.shape[1:3], -1)
+            )  # C, B, T, E
+            center_features = torch.cat((features, embedded_left_contexts_), -1)  # C, B, T, F+E
+            center_logits = self.center_encoder(center_features)  # C, B, T, F'
+            center_probs = F.log_softmax(center_logits, -1)
+            center_probs = center_probs.permute((1, 2, 3, 0))  # B, T, F', C
```
Comment (michelwi) [inline i6_models/parts/factored_hybrid/diphone.py:156]: I hope you checked that this leaves the outputs in the same order as expected in RASR diphone-dense tying?

File: i6_models/parts/factored_hybrid/diphone.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        assert self.context_mix_mlp_dim > 0
+        assert self.context_mix_mlp_num_layers > 0
+        assert self.n_contexts > 0
+        assert self.num_hmm_states_per_phone > 0
+        assert self.num_inputs > 0
+        assert self.left_context_embedding_dim > 0
+        assert 0.0 <= self.dropout <= 1.0, "dropout must be a probability"
+
+
+class DiphoneBackendV1(nn.Module):
```
Comment (michelwi) [inline i6_models/parts/factored_hybrid/diphone.py:78]: I am a bit unhappy about the name "backend" as that term says exactly nothing. I once talked to guys from Paderborn who were doing denoising and source separation of audio spectra and when they said "backend" they meant "feature extraction and other ASR model stuff" 🤷 

I was trying to find a good name that implies that here embeddings are converted to (log-)probabilities.. but I failed.. pls help :)

Comment (JackTemaki) [reply i6_models/parts/factored_hybrid/diphone.py:78]: +1 for Willis statement. Please really do not call it Backend. Maybe `DiphoneOutputBlock`?

File: i6_models/parts/factored_hybrid/diphone.py (hunk 4 above)
Comment (JackTemaki) [reply i6_models/parts/factored_hybrid/diphone.py:113]: I just want to raise that "maybe" dataclasses as output might not be optimal for some graph tracing/scripting logic from PyTorch.

File: i6_models/parts/factored_hybrid/diphone.py (hunk 7 above)
Comment (Marvin84) [reply i6_models/parts/factored_hybrid/diphone.py:78]: +1 for Moritz idea.
In defense of the naming, in the literature and commonly within ASR community, the feature extraction and all respective methods such as source separation, beam forming, and enhancement are denoted as ASR frontend and the model itself is called backend. I find the naming well chosen since it reminds the user that this can be combined with any encoder architecture or speech representation h(x).
The new generation calls the similar part of an RNN-T as the "decoder", which is certainly not a better term.

Comment (Marvin84) [reply i6_models/parts/factored_hybrid/diphone.py:78]: We just had a live discussion with @JackTemaki and @ZijYang and we think a nice name for this can be `DiphoneJoinerBlock` reminding the similar RNN-T case

File: i6_models/parts/factored_hybrid/diphone.py (hunk 3 above)
Comment (Marvin84) [reply i6_models/parts/factored_hybrid/diphone.py:163]: We typically keep separate outputs for training and combine the two outputs only during inference, meaning that we do need a loss computation also on the `left_output`

Comment (michelwi) [reply i6_models/parts/factored_hybrid/diphone.py:163]: this is an argument in favor of having different kinds of outputs depending on train yes/no i.e. using different dataclasses (or e.g. tuples if we want to follow @JackTemaki s suggestion)

File: i6_models/parts/factored_hybrid/util.py (hunk 1 above)
Comment (Marvin84) [reply i6_models/parts/factored_hybrid/util.py:7]: >  it would be related to num_hmm_states_per_phone

Yes, I also agree.

File: i6_models/parts/factored_hybrid/diphone.py (hunk 5 above)
Comment (michelwi) [reply i6_models/parts/factored_hybrid/diphone.py:60]: offline discussion: With the fear of a statefull activation module this is indeed intended as a function that creates new instances every time it is called.

File: i6_models/parts/factored_hybrid/diphone.py (hunk 3 above)
Comment (Marvin84) [reply i6_models/parts/factored_hybrid/diphone.py:163]: Another aspect to consider @NeoLegends  is whether we want to have two separate modules for the architecture that uses frame-wise cross-entropy and the one that uses sequence-level cross-entropy, i.e. fullsum. Since in the latter also training would rely on the combination of the outputs.

File: i6_models/parts/factored_hybrid/util.py (hunk 2 above)
Comment (NeoLegends) [reply i6_models/parts/factored_hybrid/util.py:42]: That one does not work with `nn.Sequential`, which is why I don't use it here.

File: i6_models/parts/factored_hybrid/diphone.py (hunk 4 above)
Comment (michelwi) [reply i6_models/parts/factored_hybrid/diphone.py:113]: Maybe a 
```
Tuple[Optional[Tensor], Optional[Tensor], Optional[Tensor], Optional[Tensor]]
```
with [joined, center, left, left_embedding] and depending on the conditions, each can be None if not applicable.?

Or should we just always return all of the above and then in training / inference some are just not used?

File: i6_models/parts/factored_hybrid/util.py (hunk 2 above)
Comment (michelwi) [reply i6_models/parts/factored_hybrid/util.py:42]: offline discussion: this does not work for nn.Sequential

File: i6_models/parts/factored_hybrid/diphone.py (hunk 4 above)
Comment (NeoLegends) [reply i6_models/parts/factored_hybrid/diphone.py:113]: I'm moving to tuples.

File: i6_models/parts/factored_hybrid/diphone.py (hunk 6 above)
Comment (NeoLegends) [reply i6_models/parts/factored_hybrid/diphone.py:156]: I'm going to verify what I'm doing here is right before this PR can be merged. Either via test or via setup.

File: i6_models/parts/factored_hybrid/diphone.py (hunk 7 above)
Comment (michelwi) [reply i6_models/parts/factored_hybrid/diphone.py:78]: I like "Joiner" better than "Backend" and "Output" most of them. Although this is not necessarily the (final) output of the model.

`DiphoneProbabilityBlock` ? or even just (`Factored`-)`DiphoneBlock` ?

File: i6_models/parts/factored_hybrid/diphone.py (hunk 3 above)
Comment (michelwi) [reply i6_models/parts/factored_hybrid/diphone.py:163]: > sequence-level cross-entropy, i.e. fullsum.

*"e.g. fullsum"

Daniel has already offered to do a pr for the LFMMI op xP

File: i6_models/parts/factored_hybrid/diphone.py (hunk 7 above)
Comment (Marvin84) [reply i6_models/parts/factored_hybrid/diphone.py:78]: I also like `FactoredDiphoneBlock`. 
BTW, one feedback I got from people outside our research group is that there is a marketing problem with the name of factored hybrid, and if we had a fancy new name people would be more open to try it out. Independently of the name, using up-to-date terminologies that do not remind the GMM-HMM systems is preferable, IMO.

File: i6_models/parts/factored_hybrid/diphone.py (hunk 3 above)
Comment (Marvin84) [reply i6_models/parts/factored_hybrid/diphone.py:163]: > how easy it is to change model classes from training to inference in torch

Let's also ask @SimBe195 and @Judyxujj, they are also running torch trainings for a while now

File: i6_models/parts/factored_hybrid/diphone.py (hunk 7 above)
Comment (michelwi) [reply i6_models/parts/factored_hybrid/diphone.py:78]: > if we had a fancy new name people would be more open to try it out. 

_seigh_

> Independently of the name, using up-to-date terminologies that do not remind the GMM-HMM systems is preferable, IMO.

Nothing inherently needs this to be for "phone"s it could also be FactoredDichar or DiBPE.. (except for the adherence to rasrs phoneme numbering scheme in `PhonemeStateClassV1`).. 

So should we drop the "Phone" branding?^^

File: i6_models/parts/factored_hybrid/diphone.py (hunk 3 above)
Comment (JackTemaki) [reply i6_models/parts/factored_hybrid/diphone.py:163]: I would not do it in separate modules, so that you always load everything consistently, but you can also not use `self.training`. What you can have is separate functions. `self.training` has the problem that it is really for "training" only, not e.g. for cross validation during training. We had bugs in setups before where we then applied dropout and specaugment for the validation sets.

So when you use `train_step` or `forward_step / onnx_export` you can call two different functions

File: i6_models/parts/factored_hybrid/diphone.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        :return: tuple of logits for p(c|l,x), p(l|x) and the embedded left context values.
+        """
+
+        left_logits = self.left_context_encoder(features)  # B, T, C
+        # in training we forward exactly one context per T, so: B, T, E
+        contexts_embedded_left = self.left_context_embedding(contexts_left)
+        features_center = torch.cat((features, contexts_embedded_left), -1)  # B, T, F+E
+        logits_center = self.center_encoder(features_center)  # B, T, C
+
+        return logits_center, left_logits, contexts_embedded_left
```
Comment (michelwi) [inline i6_models/parts/factored_hybrid/diphone.py:110]: `logits_center` vs `left_logits` 
consistent order of "position" and "what is it"  would be appreciated

Comment (NeoLegends) [reply i6_models/parts/factored_hybrid/diphone.py:110]: Ah, yes.

File: i6_models/parts/factored_hybrid/diphone.py (hunk 6 above)
Comment (NeoLegends) [reply i6_models/parts/factored_hybrid/diphone.py]: I have decoding results now with reasonable WERs and these verify that what we're doing here is right. :)

File: i6_models/parts/factored_hybrid/diphone.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+        context_mix_mlp_dim: inner dimension of the context mixing MLP layers
+        context_mix_mlp_num_layers: how many hidden layers on the MLPs there should be
+        dropout: dropout probabilty
+        left_context_embedding_dim: embedding dimension of the left context
+            values. Good choice is in the order of num_contexts.
+        num_contexts: the number of raw phonemes/acoustic contexts
+        num_hmm_states_per_phone: the number of HMM states per phoneme
+        num_inputs: input dimension of the output block, must match w/ output dimension
+            of main encoder (e.g. Conformer)
+        phoneme_state_class: the phoneme state augmentation to apply
```
Comment (michelwi) [inline i6_models/parts/factored_hybrid/diphone.py:31]:
```suggestion
        boundary_class: the phoneme state augmentation to apply
```

also make a conscious decision on the order of parameters and then order the docstring in the same order as the parameters defined pls.

File: i6_models/parts/factored_hybrid/diphone.py [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+            num_layers=cfg.context_mix_mlp_num_layers,
+            dropout=cfg.dropout,
+            activation=cfg.activation,
+        )
+        self.left_context_embedding = nn.Embedding(cfg.num_contexts, cfg.left_context_embedding_dim)
+        self.center_encoder = get_mlp(
+            num_input=cfg.num_inputs + cfg.left_context_embedding_dim,
+            num_output=self.num_center,
+            hidden_dim=cfg.context_mix_mlp_dim,
+            num_layers=cfg.context_mix_mlp_num_layers,
```
Comment (michelwi) [inline i6_models/parts/factored_hybrid/diphone.py:89]: Is this intentionally using the same dim/num_layers as the (left-)context MLP? Maybe rename those parameters to a name that reflects that both left-context and center have these dims?

Comment (NeoLegends) [reply i6_models/parts/factored_hybrid/diphone.py:89]: My intuition says the context mixing MLPs themselves don't have a big impact on model performance as long as they're sized sufficiently to perform their context mixing job, but maybe for the higher order contexts the size becomes more relevant due to the additional dependencies. Then again, this is also just intuition.

@Marvin84 Do you have any experience about the difference the sizing of these layers makes on the overall WER? The dim and number of layers here are the same because (I think) we never really experimented around much with them.

Comment (NeoLegends) [reply i6_models/parts/factored_hybrid/diphone.py:94]: It would certainly help if it turned out that these layers could even be sized down, and thereby speed up the score computation vs. a CART model (joint FH diphone is slower than triphone CART due to batching the context mixing MLPs over all contexts, whereas the CART just upsizes the model output via one linear layer to the final output dim).

Comment (Marvin84) [reply i6_models/parts/factored_hybrid/diphone.py:94]: I experimented with BLSTM based models, and in that case one layer against two layers had max `0.2%` absolute WER degradation on Hub500 (trained on SWB300h). Moreover, lower embed dimension could lead to absolute `0.1%` WER degradation. 
However, I noticed that for from-scratch Conformed based alignment models (monophone PHMM) having or not having MLP layers matter a lot. I have not checked the effect of number of MLP layers for Viterbi training with fixed path, though.

> speed up the score computation vs. a CART model 

How much speedup do you expect to get?

Comment (NeoLegends) [reply i6_models/parts/factored_hybrid/diphone.py:94]: image

Upper line is triphone CART, lower line diphone FH. Would be nice if the TF cost for both was at least equal. It's unlikely we can get FH any better, though.

Comment (michelwi) [reply i6_models/parts/factored_hybrid/diphone.py:94]: > Upper line is triphone CART, lower line diphone FH.

should we not compare explicit full-diphone to joint forwarding of factored diphone?

Comment (NeoLegends) [reply i6_models/parts/factored_hybrid/diphone.py:94]: In any case I think making this any more configurable can be postponed to a V2.

> should we not compare explicit full-diphone to joint forwarding of factored diphone?

Technically, yes. There won't be a noticeable difference to a triphone CART model wrt. the NN cost, however, as the size of the output layers is almost equal (at least for Librispeech, where it's 12k CART vs 11.5k diphone).

**Review by michelwi — APPROVED** [review]

**Review by Marvin84 — APPROVED** [review]
